### PR TITLE
#41 get data from tables

### PR DESCRIPTION
**Pull request by AVM82**
Connect to #41

### CODE REVIEW

File: src/main/java/juja/sqlcmd/DataSet.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package juja.sqlcmd;
+
+import java.util.Arrays;
+
+public class DataSet {
+
+    private String[] oneRow;
```
Comment (ivshapovalov) [inline src/main/java/juja/sqlcmd/DataSet.java]: Is it essential to specify number in varibale name?

File: src/main/java/juja/sqlcmd/DatabaseManager.java [hunk 2]
```diff
@@ -51,15 +51,58 @@ public boolean connect(String database, String user, String password) {
...
             return resultSet.getInt(1);
         }
     }
 
-    public void closeConnection() {
+    public DataSet[] getTableData(String tableName) {
+        String query = String.format("SELECT * FROM \"%s\"", tableName);
+        DataSet[] tableData = new DataSet[0];
+        if (connection == null) {
+            return tableData;
```
Comment (ivshapovalov) [inline src/main/java/juja/sqlcmd/DatabaseManager.java]: See that materials about one/many return from methods in Java
https://stackoverflow.com/questions/15527482/is-using-return-at-the-start-of-method-bad-coding-practice
https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement
and this one
https://dzone.com/articles/common-code-violations-java

File: src/main/java/juja/sqlcmd/DatabaseManager.java [hunk 3]
```diff
@@ -51,15 +51,58 @@ public boolean connect(String database, String user, String password) {
...
+        }
+        if (!isTableExist(tableName)) {
```
Comment (ivshapovalov) [inline src/main/java/juja/sqlcmd/DatabaseManager.java]: Exists probably?

**Review by ivshapovalov — CHANGES_REQUESTED** [review]

File: src/main/java/juja/sqlcmd/DatabaseManager.java (hunk 3 above)
Comment (AVM82) [reply src/main/java/juja/sqlcmd/DatabaseManager.java]: I will renaming this method.
Thank

File: src/main/java/juja/sqlcmd/DataSet.java (hunk 1 above)
Comment (AVM82) [reply src/main/java/juja/sqlcmd/DataSet.java]: I will renaming this varibale.

File: src/main/java/juja/sqlcmd/DatabaseManager.java (hunk 2 above)
Comment (AVM82) [reply src/main/java/juja/sqlcmd/DatabaseManager.java]: I tried to change the logic. @ivshapovalov  please look it.

File: src/main/java/juja/sqlcmd/DatabaseManager.java [hunk 4]
```diff
@@ -51,15 +51,54 @@ public boolean connect(String database, String user, String password) {
...
+        }
+        return new DataSet[0];
+    }
+
+    private boolean hasTable(String tableName) {
+        String query = String.format("SELECT to_regclass('%s')", tableName);
+        try (Statement statement = connection.createStatement();
+             ResultSet resultSet = statement.executeQuery(query)) {
+            resultSet.next();
+            return resultSet.getString(1) != null;
```
Comment (Shtramak) [inline src/main/java/juja/sqlcmd/DatabaseManager.java]: Here you can use `return resultSet.next();`. That statement will give you an answer if the result set is not emtpy

File: src/main/java/juja/sqlcmd/DatabaseManager.java [hunk 5]
```diff
@@ -51,15 +51,54 @@ public boolean connect(String database, String user, String password) {
...
+        } catch (SQLException e) {
+            System.out.println(e.getMessage());
+        }
+        return false;
+    }
+
+    public void close() {
         try {
-            connection.close();
+            if (connection != null && !connection.isClosed()) {
```
Comment (Shtramak) [inline src/main/java/juja/sqlcmd/DatabaseManager.java:101]: connection.isClosed() is redutant. If the connection is alredy closed than `connection.close();` will do nothing

File: src/main/java/juja/sqlcmd/DatabaseManager.java [hunk 6]
```diff
@@ -51,15 +51,54 @@ public boolean connect(String database, String user, String password) {
...
+    }
+
+    private boolean hasTable(String tableName) {
+        String query = String.format("SELECT to_regclass('%s')", tableName);
+        try (Statement statement = connection.createStatement();
+             ResultSet resultSet = statement.executeQuery(query)) {
+            resultSet.next();
+            return resultSet.getString(1) != null;
+        } catch (SQLException e) {
+            System.out.println(e.getMessage());
```
Comment (Shtramak) [inline src/main/java/juja/sqlcmd/DatabaseManager.java]: Why you are handling exceptions this way? This is not a handling, just printing `e.getMessage()` to console. There is some approcah "Throw early, catch late", here you can find more explanations - http://bit.ly/2Bo9zQP. It's better to rethrow SQLException in this case. But if you still would like to log somehow ( like sout() ) the exception message you can do it this way:
```
} catch (SQLException e) {
    System.out.println(e.getMessage());
    throw e;
}
```

File: src/test/java/juja/sqlcmd/DatabaseManagerTest.java [hunk 7]
```diff
@@ -48,65 +65,109 @@ public void testConnectionIfWrongPassword() {
...
+        return oneRow;
     }
 
-    private void preparationEnvironment() throws ClassNotFoundException, SQLException {
-        this.dbName = new Date().toString();
-        Connection testConnection = getTestConnection();
-        try (Statement statement = testConnection.createStatement()) {
-            statement.execute(String.format("CREATE DATABASE \"%s\" ", this.dbName));
-            testConnection.close();
+    @After
```
Comment (Shtramak) [inline src/test/java/juja/sqlcmd/DatabaseManagerTest.java:140]: I think that @After and @AfterClass blocks should  go after @Before and @BeforeClass. With this blocks you describe a part of test-environment logic and methods annotating with @After, @AfterClass, @Before and @BeforeClass have a strong relation between each other

File: src/test/java/juja/sqlcmd/DatabaseManagerTest.java [hunk 8]
```diff
@@ -19,8 +21,23 @@
     private static final String USER = "sqlcmd";
     private static final String PASSWORD = "sqlcmd";
     private static final String TEST_DB_CONNECTION_URL = "jdbc:postgresql://localhost:5432/";
+
+    private static Connection connection;
+
     private DatabaseManager databaseManager;
-    private String dbName;
+    private static String dbName;
```
Comment (Shtramak) [inline src/test/java/juja/sqlcmd/DatabaseManagerTest.java]: This line should be in the same block with `private static Connection connection;`

File: src/main/java/juja/sqlcmd/DatabaseManager.java (hunk 5 above)
Comment (AVM82) [reply src/main/java/juja/sqlcmd/DatabaseManager.java:101]: I would like to discuss this

File: src/test/java/juja/sqlcmd/DatabaseManagerTest.java (hunk 8 above)
Comment (AVM82) [reply src/test/java/juja/sqlcmd/DatabaseManagerTest.java]: Ok, i will fix this

File: src/test/java/juja/sqlcmd/DatabaseManagerTest.java (hunk 7 above)
Comment (AVM82) [reply src/test/java/juja/sqlcmd/DatabaseManagerTest.java:140]: Thanks for your comment, maybe you're right.
I'll think about it

**Comment by adaleorion** [comment]
@AVM82 do you have change for this pull request ?

File: src/main/java/juja/sqlcmd/DatabaseManager.java (hunk 4 above)
Comment (AVM82) [reply src/main/java/juja/sqlcmd/DatabaseManager.java]: Thank you for your comment.
I will do as you advised.

**Comment by AVM82** [comment]
@danilkuznetsov No More changes I would like to do for this pull request.

**Review by kortov — APPROVED** [review]

**Comment by adaleorion** [comment]
@ivshapovalov if you agree, please approve this pull request

**Review by adaleorion — APPROVED** [review]
good

**Review by Shtramak — APPROVED** [review]

**Review by ivshapovalov — APPROVED** [review]

**Pull request merged**